Kernel/Arch/x86/common/Boot/boot.S

```diff
@@ -307,14 +310,34 @@ real_start:
     hlt
 
 kernel_not_too_large:
-
-
-#if ARCH(X86_64)
-    /* test for long mode presence, save the most important registers from corruption */
+    /* test for PAE presence, save the most important registers from corruption */
     pushl %eax
     pushl %edx
     pushl %ebx
 
+    movl $0x1, %eax       /* PAE presence is in CPUID input 0x1 */
+    cpuid
+    testl $(1 << 6), %edx /* Test if the PAE-bit, which is bit 6, is set in the edx register. */
+    jnz pae_supported     /* If the bit is not set, there is no PAE capability. */
+
+    /*  from now on, we don't really care about booting because PAE is not supported.
+         the flow from now is like so:
+         1. Copy all necessary parts to low memory section in RAM
+         2. Jump to that section
+         3. In that section we do:
+            a. exit protected mode to pure 16 bit real mode
+            b. load the "PAE is not supported" string, call the BIOS print to screen service
+            c. halt
+    */
```

Comment:
Maybe instead of duplicating this rather long comment we should move it to `print_and_halt` instead.